modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -52,18 +54,36 @@ def wait_computations_if_benchmark_mode(func):
     """
     if BenchmarkMode.get():
 
+        def extract_waitable(obj):
+            """Return an iterable of waitable objects contained in the `obj`."""
+            if hasattr(obj, "wait"):
+                return (obj,)
+            # fastpath for NumPy array of partitions
+            elif (
+                isinstance(obj, np.ndarray)
+                and obj.dtype == np.object
+                and len(obj) > 0
+                and hasattr(obj[np.unravel_index(0, shape=obj.shape)], "wait")
+            ):
+                return obj.flatten()
+            elif is_list_like(obj):
+                return tuple(chain(*(extract_waitable(o) for o in obj)))
+            else:
+                return tuple()
+
         @wraps(func)
         def wait(*args, **kwargs):
             """Wait for computation results."""
             result = func(*args, **kwargs)
-            if isinstance(result, tuple):
-                partitions = result[0]
-            else:
-                partitions = result
+            partitions = extract_waitable(result)
+            ErrorMessage.catch_bugs_and_request_email(
+                failure_condition=len(partitions) == 0,
+                extra_log=f"No partition to wait was returned from the wrapped function: {func}",
+            )
             # need to go through all the values of the map iterator
             # since `wait` does not return anything, we need to explicitly add
             # the return `True` value from the lambda
-            all(map(lambda partition: partition.wait() or True, partitions.flatten()))
+            all(map(lambda partition: partition.wait() or True, partitions))
```

Comment:
While you're modifying this line, could you please simplify the code to 

```suggestion
            [partition.wait() for partition in partitions]
```

I think the existing line is effectively doing the same thing.